File: Programmers/[60061] 기둥과 보 설치/KingDonggyu/60061.js
function solution(n, build_frame) {
  const wall = Array(n + 1).fill(0)
    .map((_) => Array(n + 1).fill(0)
    .map((_) => [false, false]));

  const checkPillar = (x, y) => {
    // 바닥 위
    if (!y) {
      return true;
    }
    // 기둥 위
    if (y > 0 && wall[x][y - 1][0]) {
      return true;
    }
    // 보 위
    if (
      wall[x][y][1] || 
      (x > 0 && wall[x - 1][y][1])
    ) {
      return true;
    }
  };

  const checkBeam = (x, y) => {
    // 기둥 위
    if (
      (y > 0 && wall[x][y - 1][0]) || 
      (x < n && wall[x + 1][y - 1][0])
    ) {
      return true;
    }
    // 보 사이
    if (
      x > 0 && wall[x - 1][y][1] && 
      x < n && wall[x + 1][y][1]
    ) {
      return true;
    }
  };

  const checkDelete = () => {
    for (let x = 0; x < n + 1; x++) {
      for (let y = 0; y < n + 1; y++) {
        if (wall[x][y][0] && !checkPillar(x, y)) {
          return false;
        }
        if (wall[x][y][1] && !checkBeam(x, y)) {
          return false;
        }
      }
    }
    return true;
  };

  build_frame.forEach(([x, y, a, b]) => {
    if (!b) {
      wall[x][y][a] = false;
      if (!checkDelete()) {
        wall[x][y][a] = true;
      }
      return;
    }

    const install = a ? checkBeam(x, y) : checkPillar(x, y);

    if (install) {
      wall[x][y][a] = true;
    }
  });

  const answer = [];

  for (let x = 0; x < n + 1; x++) {
    for (let y = 0; y < n + 1; y++) {
      wall[x][y][0] && answer.push([x, y, 0]);
      wall[x][y][1] && answer.push([x, y, 1]);
    }
  }

  return answer;
}

File: Programmers/[60061] 기둥과 보 설치/JFe/60061.py
def check(answer):
    for x, y, a in answer:
        # 기둥인 경우
        if a == 0:
            # 바닥 위 or 보의 한쪽 끝 부분 위 or 다른 기둥 위인지 확인
            if y == 0 or [x-1, y, 1] in answer or [x, y, 1] in answer or [x, y-1, 0] in answer: continue
            return False
        # 보인 경우
        if a == 1:
            # 한쪽 끝 부분이 기둥 위 or 양쪽 끝 부분이 다른 보와 동시에 연결 상태인지 확인
            if [x, y-1, 0] in answer or [x+1, y-1, 0] in answer or [x-1, y, 1] in answer and [x+1, y, 1] in answer: continue
            return False
    return True
    
def solution(n, build_frame):
    answer = []
    for x, y, a, b in build_frame:
        # 설치하는 경우
        if b == 1:
            answer.append([x, y, a])
            if not check(answer):
                answer.remove([x, y, a])
        # 삭제하는 경우
        elif b == 0:
            answer.remove([x, y, a])
            if not check(answer):
                answer.append([x, y, a])
    answer.sort(key = lambda x : (x[0], x[1], x[2]))
    return answer